core/block_validator.go

```diff
@@ -55,37 +55,50 @@ func NewBlockValidator(config *params.ChainConfig, headerChain *HeaderChain, eng
 func (v *BlockValidator) ValidateBody(block *types.Block) error {
 	nodeCtx := common.NodeLocation.Context()
 	// Check whether the block's known, and if not, that it's linkable
-	if v.hc.bc.processor.HasBlockAndState(block.Hash(), block.NumberU64()) {
-		return ErrKnownBlock
+	if nodeCtx == common.ZONE_CTX {
```

Comment:
I think this escape applies in region and prime context too still, no?

Reply:
Prime and region don't have a state root so calling this function would technically be wrong

Reply:
Then we should just call `HasBlock()` right?

Reply:
HasBlock check has no meaning, in ValidateBody right?

Reply:
ok, I see, just remove the check then

Reply:
ah right, since we write all blocks to the db, not just canonical blocks. Got it